File: apps/backend/src/pbdh_backend/contracts/resource_package.py
import copy
import hashlib
import json
import struct
from collections.abc import Mapping
from typing import Any


FAMILY = "resource-package"
LEGACY_VERSION = "1.0.0-alpha.1"
DIGEST_DOMAIN = "pbdh-resource-package-digest-v1"


def _utf16_sort_key(value: str) -> bytes:
    return value.encode("utf-16-be")


def _canonicalize(value: Any) -> str:
    if value is None:
        return "null"
    if isinstance(value, bool):
        return "true" if value else "false"
    if isinstance(value, str):
        value.encode("utf-8")
        return json.dumps(value, ensure_ascii=False, separators=(",", ":"))
    if isinstance(value, list):
        return f"[{','.join(_canonicalize(item) for item in value)}]"
    if isinstance(value, Mapping):
        keys = sorted(value, key=_utf16_sort_key)
        return "{" + ",".join(
            f"{json.dumps(key, ensure_ascii=False)}:{_canonicalize(value[key])}"
            for key in keys
        ) + "}"
    raise ValueError("Resource Package canonical JSON supports only string, boolean, null, array, and object")


def _normalized_document(document: Mapping[str, Any]) -> dict[str, Any]:
    content = copy.deepcopy(dict(document))
    content.pop("snapshotDigest", None)
    content["targets"].sort(
        key=lambda item: (item["systemPackageId"], item["version"])
    )
    content["assets"].sort(key=lambda item: item["id"])
    content["resources"].sort(key=lambda item: (item["path"], item["id"]))
    for resource in content["resources"]:
        resource.get("replacements", []).sort(
            key=lambda item: (item["replacementId"], item["targetResourceId"])
        )
    content["emptyDirectories"].sort()
    return content


def _frame(frame_type: str, payload: bytes) -> bytes:
    type_bytes = frame_type.encode("utf-8")
    return (
        struct.pack(">I", len(type_bytes))
        + type_bytes
        + struct.pack(">Q", len(payload))
        + payload
    )


def compute_resource_package_snapshot_digest(
    document: Mapping[str, Any],
    media: Mapping[str, bytes],
) -> str:
    parts = [
        _frame("domain", DIGEST_DOMAIN.encode("utf-8")),
        _frame(
            "logical-document",
            _canonicalize(_normalized_document(document)).encode("utf-8"),
        ),
    ]
    for asset in sorted(document["assets"], key=lambda item: item["id"]):
        asset_id = asset["id"]
        if asset_id not in media:
            raise ValueError(f"Missing media bytes: {asset_id}")
        parts.append(_frame("asset-id", asset_id.encode("utf-8")))
        parts.append(_frame("asset-bytes", media[asset_id]))
    return f"sha256:{hashlib.sha256(b''.join(parts)).hexdigest()}"


def _diagnostic(
    code: str,
    location: str,
    params: Mapping[str, Any],
    version: str,
) -> dict[str, Any]:
    return {
        "code": code,
        "severity": "error",
        "family": FAMILY,
        "version": version,
        "location": location,
        "params": dict(params),
    }


def _pointer_segment(value: str) -> str:
    return value.replace("~", "~0").replace("/", "~1")


def validate_resource_package_semantics(
    document: Mapping[str, Any],
    media: Mapping[str, bytes],
) -> list[dict[str, Any]]:
    diagnostics: list[dict[str, Any]] = []
    version = str(document.get("contractVersion", LEGACY_VERSION))
    resource_ids: set[str] = set()
    targets: set[tuple[str, str]] = set()
    asset_ids: set[str] = set()

    for index, resource in enumerate(document["resources"]):
        resource_id = resource["id"]
        if resource_id in resource_ids:
            diagnostics.append(_diagnostic(
                "resource-package.resource-id.duplicate",
                f"/resources/{index}/id",
                {"id": resource_id},
                version,
            ))
        resource_ids.add(resource_id)

    for index, target in enumerate(document["targets"]):
        key = (target["systemPackageId"], target["version"])
        if key in targets:
            diagnostics.append(_diagnostic(
                "resource-package.target.duplicate",
                f"/targets/{index}",
                {
                    "systemPackageId": target["systemPackageId"],
                    "version": target["version"],
                },
                version,
            ))
        targets.add(key)

    for index, asset in enumerate(document["assets"]):
        asset_id = asset["id"]
        if asset_id in asset_ids:
            diagnostics.append(_diagnostic(
                "resource-package.asset-id.duplicate",
                f"/assets/{index}/id",
                {"id": asset_id},
                version,
            ))
        asset_ids.add(asset_id)
        media_bytes = media.get(asset_id)
        if media_bytes is None:
            diagnostics.append(_diagnostic(
                "resource-package.media.bytes-missing",
                f"/assets/{index}/id",
                {"assetId": asset_id},
                version,
            ))
            continue
        actual_id = f"sha256:{hashlib.sha256(media_bytes).hexdigest()}"
        if actual_id != asset_id:
            diagnostics.append(_diagnostic(
                "resource-package.media.digest-mismatch",
                f"/assets/{index}/id",
                {"actual": actual_id, "expected": asset_id},
                version,
            ))
        if str(len(media_bytes)) != asset["byteLength"]:
            diagnostics.append(_diagnostic(
                "resource-package.media.byte-length-mismatch",
                f"/assets/{index}/byteLength",
                {"actual": str(len(media_bytes)), "expected": asset["byteLength"]},
                version,
            ))

    for resource_index, resource in enumerate(document["resources"]):
        replacement_ids: set[str] = set()
        for replacement_index, replacement in enumerate(resource.get("replacements", [])):
            replacement_id = replacement["replacementId"]
            if replacement_id in replacement_ids:
                diagnostics.append(_diagnostic(
                    "resource-package.replacement-id.duplicate",
                    f"/resources/{resource_index}/replacements/{replacement_index}/replacementId",
                    {"replacementId": replacement_id},
                    version,
                ))
            replacement_ids.add(replacement_id)
            target_resource_id = replacement["targetResourceId"]
            if target_resource_id not in resource_ids:
                diagnostics.append(_diagnostic(
                    "resource-package.replacement-target.missing",
                    f"/resources/{resource_index}/replacements/{replacement_index}/targetResourceId",
                    {"targetResourceId": target_resource_id},
                    version,
                ))
        for slot, asset_id in resource["media"].items():
            if asset_id not in asset_ids:
                diagnostics.append(_diagnostic(
                    "resource-package.media.asset-undeclared",
                    f"/resources/{resource_index}/media/{_pointer_segment(slot)}",
                    {"assetId": asset_id},
                    version,
                ))

    if not diagnostics:
        actual = compute_resource_package_snapshot_digest(document, media)
        if actual != document["snapshotDigest"]:
            diagnostics.append(_diagnostic(
                "resource-package.snapshot-digest.mismatch",
                "/snapshotDigest",
                {"actual": actual, "expected": document["snapshotDigest"]},
                version,
            ))

    return sorted(diagnostics, key=lambda item: (item["location"], item["code"]))

File: packages/contract-runtime/src/resource-package.ts
import type { ContractDiagnostic } from "./index.ts";

const FAMILY = "resource-package";
export const RESOURCE_PACKAGE_VERSION = "1.0.0";
export const LEGACY_RESOURCE_PACKAGE_VERSION = "1.0.0-alpha.1";
const DIGEST_DOMAIN = "pbdh-resource-package-digest-v1";
const VERSION_CONTENT_DOMAIN = "pbdh-resource-package-version-content-v1";

type JsonValue = null | boolean | string | JsonValue[] | { [key: string]: JsonValue };

export type ResourceReplacement = {
  replacementId: string;
  targetResourceId: string;
};

export type ResourcePackageLogicalDocument = {
  contractVersion: typeof RESOURCE_PACKAGE_VERSION | typeof LEGACY_RESOURCE_PACKAGE_VERSION;
  package: {
    id: string;
    version: string;
    name: string;
    description: string;
  };
  targets: Array<{ systemPackageId: string; version: string }>;
  license: { label: string; declaration: string };
  publication?: {
    language: string;
    tags: string[];
    coverAssetId: string;
  };
  forkSource: JsonValue;
  assets: Array<{
    id: string;
    mediaType: "image/webp";
    byteLength: string;
    width: string;
    height: string;
  }>;
  resources: Array<{
    id: string;
    path: string;
    template: { id: string; version: string };
    presentation: ResourcePresentation;
    data: JsonValue;
    replacements?: ResourceReplacement[];
    media: Record<string, string>;
  }>;
  emptyDirectories: string[];
  snapshotDigest: string;
};

export type ResourcePresentation = {
  width: string;
  height: string;
  unit: "mm";
  mode: "text" | "split" | "image";
  fixedRatio: boolean;
};

export type ResourcePackageMedia = ReadonlyMap<string, Uint8Array>;

function compareCodePoints(left: string, right: string): number {
  const leftPoints = Array.from(left, (character) => character.codePointAt(0)!);
  const rightPoints = Array.from(right, (character) => character.codePointAt(0)!);
  const length = Math.min(leftPoints.length, rightPoints.length);
  for (let index = 0; index < length; index += 1) {
    const difference = leftPoints[index]! - rightPoints[index]!;
    if (difference !== 0) return difference;
  }
  return leftPoints.length - rightPoints.length;
}

function assertValidUnicode(value: string): void {
  for (let index = 0; index < value.length; index += 1) {
    const unit = value.charCodeAt(index);
    if (unit >= 0xd800 && unit <= 0xdbff) {
      const next = value.charCodeAt(index + 1);
      if (next < 0xdc00 || next > 0xdfff) throw new Error("Invalid Unicode surrogate");
      index += 1;
    } else if (unit >= 0xdc00 && unit <= 0xdfff) {
      throw new Error("Invalid Unicode surrogate");
    }
  }
}

function canonicalize(value: JsonValue): string {
  if (value === null) return "null";
  if (typeof value === "string") {
    assertValidUnicode(value);
    return JSON.stringify(value);
  }
  if (typeof value === "boolean") return value ? "true" : "false";
  if (Array.isArray(value)) return `[${value.map(canonicalize).join(",")}]`;

  const keys = Object.keys(value).sort();
  for (const key of keys) assertValidUnicode(key);
  return `{${keys
    .map((key) => `${JSON.stringify(key)}:${canonicalize(value[key]!)}`)
    .join(",")}}`;
}

function normalizedDocument(document: ResourcePackageLogicalDocument): JsonValue {
  const { snapshotDigest: _snapshotDigest, ...content } = structuredClone(document);
  content.targets.sort((left, right) => {
    const idOrder = compareCodePoints(left.systemPackageId, right.systemPackageId);
    return idOrder || compareCodePoints(left.version, right.version);
  });
  content.assets.sort((left, right) => compareCodePoints(left.id, right.id));
  content.resources.sort((left, right) => {
    const pathOrder = compareCodePoints(left.path, right.path);
    return pathOrder || compareCodePoints(left.id, right.id);
  });
  content.resources.forEach((resource) => resource.replacements?.sort((left, right) => {
    const idOrder = compareCodePoints(left.replacementId, right.replacementId);
    return idOrder || compareCodePoints(left.targetResourceId, right.targetResourceId);
  }));
  content.emptyDirectories.sort(compareCodePoints);
  return content as JsonValue;
}

function normalizedVersionContent(document: ResourcePackageLogicalDocument): JsonValue {
  const content = normalizedDocument(document) as Omit<ResourcePackageLogicalDocument, "snapshotDigest">;
  const { version: _version, ...packageIdentity } = content.package;
  content.package = packageIdentity as ResourcePackageLogicalDocument["package"];
  return content as JsonValue;
}

function encodeFrame(type: string, payload: Uint8Array): Uint8Array {
  const encoder = new TextEncoder();
  const typeBytes = encoder.encode(type);
  const frame = new Uint8Array(4 + typeBytes.length + 8 + payload.length);
  const view = new DataView(frame.buffer);
  view.setUint32(0, typeBytes.length, false);
  frame.set(typeBytes, 4);
  view.setBigUint64(4 + typeBytes.length, BigInt(payload.length), false);
  frame.set(payload, 4 + typeBytes.length + 8);
  return frame;
}

function concatenate(parts: Uint8Array[]): Uint8Array {
  const result = new Uint8Array(parts.reduce((total, part) => total + part.length, 0));
  let offset = 0;
  for (const part of parts) {
    result.set(part, offset);
    offset += part.length;
  }
  return result;
}

function toHex(bytes: ArrayBuffer): string {
  return Array.from(new Uint8Array(bytes), (value) => value.toString(16).padStart(2, "0")).join("");
}

async function sha256(bytes: Uint8Array): Promise<string> {
  const input = new Uint8Array(bytes.byteLength);
  input.set(bytes);
  return toHex(await crypto.subtle.digest("SHA-256", input.buffer));
}

export async function computeResourcePackageSnapshotDigest(
  document: ResourcePackageLogicalDocument,
  media: ResourcePackageMedia,
): Promise<string> {
  const encoder = new TextEncoder();
  const parts = [
    encodeFrame("domain", encoder.encode(DIGEST_DOMAIN)),
    encodeFrame("logical-document", encoder.encode(canonicalize(normalizedDocument(document)))),
  ];

  for (const asset of [...document.assets].sort((left, right) => compareCodePoints(left.id, right.id))) {
    const bytes = media.get(asset.id);
    if (!bytes) throw new Error(`Missing media bytes: ${asset.id}`);
    parts.push(encodeFrame("asset-id", encoder.encode(asset.id)));
    parts.push(encodeFrame("asset-bytes", bytes));
  }

  return `sha256:${await sha256(concatenate(parts))}`;
}

export async function computeResourcePackageVersionContentHash(
  document: ResourcePackageLogicalDocument,
): Promise<string> {
  const encoder = new TextEncoder();
  return `sha256:${await sha256(concatenate([
    encodeFrame("domain", encoder.encode(VERSION_CONTENT_DOMAIN)),
    encodeFrame("logical-content", encoder.encode(canonicalize(normalizedVersionContent(document)))),
  ]))}`;
}

function semanticDiagnostic(
  version: ResourcePackageLogicalDocument["contractVersion"],
  code: string,
  location: string,
  params: Record<string, unknown>,
): ContractDiagnostic {
  return { code, severity: "error", family: FAMILY, version, location, params };
}

function sortDiagnostics(diagnostics: ContractDiagnostic[]): ContractDiagnostic[] {
  return diagnostics.sort((left, right) => {
    const locationOrder = compareCodePoints(left.location, right.location);
    if (locationOrder !== 0) return locationOrder;
    return compareCodePoints(left.code, right.code);
  });
}

export async function validateResourcePackageSemantics(
  document: ResourcePackageLogicalDocument,
  media: ResourcePackageMedia,
): Promise<ContractDiagnostic[]> {
  const diagnostics: ContractDiagnostic[] = [];
  const resourceIds = new Set<string>();
  const targets = new Set<string>();
  const assetIds = new Set<string>();

  document.resources.forEach((resource, index) => {
    if (resourceIds.has(resource.id)) {
      diagnostics.push(semanticDiagnostic(document.contractVersion,
        "resource-package.resource-id.duplicate",
        `/resources/${index}/id`,
        { id: resource.id },
      ));
    }
    resourceIds.add(resource.id);
  });

  document.targets.forEach((target, index) => {
    const key = `${target.systemPackageId}@${target.version}`;
    if (targets.has(key)) {
      diagnostics.push(semanticDiagnostic(document.contractVersion,
        "resource-package.target.duplicate",
        `/targets/${index}`,
        { systemPackageId: target.systemPackageId, version: target.version },
      ));
    }
    targets.add(key);
  });

  for (const [index, asset] of document.assets.entries()) {
    if (assetIds.has(asset.id)) {
      diagnostics.push(semanticDiagnostic(document.contractVersion,
        "resource-package.asset-id.duplicate",
        `/assets/${index}/id`,
        { id: asset.id },
      ));
    }
    assetIds.add(asset.id);
    const bytes = media.get(asset.id);
    if (!bytes) {
      diagnostics.push(semanticDiagnostic(document.contractVersion,
        "resource-package.media.bytes-missing",
        `/assets/${index}/id`,
        { assetId: asset.id },
      ));
      continue;
    }
    const actualId = `sha256:${await sha256(bytes)}`;
    if (actualId !== asset.id) {
      diagnostics.push(semanticDiagnostic(document.contractVersion,
        "resource-package.media.digest-mismatch",
        `/assets/${index}/id`,
        { actual: actualId, expected: asset.id },
      ));
    }
    if (String(bytes.byteLength) !== asset.byteLength) {
      diagnostics.push(semanticDiagnostic(document.contractVersion,
        "resource-package.media.byte-length-mismatch",
        `/assets/${index}/byteLength`,
        { actual: String(bytes.byteLength), expected: asset.byteLength },
      ));
    }
  }

  if (document.publication && !assetIds.has(document.publication.coverAssetId)) {
    diagnostics.push(semanticDiagnostic(document.contractVersion,
      "resource-package.publication-cover.asset-undeclared",
      "/publication/coverAssetId",
      { assetId: document.publication.coverAssetId },
    ));
  }

  document.resources.forEach((resource, resourceIndex) => {
    const replacementIds = new Set<string>();
    for (const [replacementIndex, replacement] of (resource.replacements ?? []).entries()) {
      if (replacementIds.has(replacement.replacementId)) {
        diagnostics.push(semanticDiagnostic(document.contractVersion,
          "resource-package.replacement-id.duplicate",
          `/resources/${resourceIndex}/replacements/${replacementIndex}/replacementId`,
          { replacementId: replacement.replacementId },
        ));
      }
      replacementIds.add(replacement.replacementId);
      if (!resourceIds.has(replacement.targetResourceId)) {
        diagnostics.push(semanticDiagnostic(document.contractVersion,
          "resource-package.replacement-target.missing",
          `/resources/${resourceIndex}/replacements/${replacementIndex}/targetResourceId`,
          { targetResourceId: replacement.targetResourceId },
        ));
      }
    }
    for (const [slot, assetId] of Object.entries(resource.media)) {
      if (!assetIds.has(assetId)) {
        diagnostics.push(semanticDiagnostic(document.contractVersion,
          "resource-package.media.asset-undeclared",
          `/resources/${resourceIndex}/media/${slot.replaceAll("~", "~0").replaceAll("/", "~1")}`,
          { assetId },
        ));
      }
    }
  });

  if (diagnostics.length === 0) {
    const actual = await computeResourcePackageSnapshotDigest(document, media);
    if (actual !== document.snapshotDigest) {
      diagnostics.push(semanticDiagnostic(document.contractVersion,
        "resource-package.snapshot-digest.mismatch",
        "/snapshotDigest",
        { actual, expected: document.snapshotDigest },
      ));
    }
  }

  return sortDiagnostics(diagnostics);
}
